qiskit/circuit/library/standard_gates/x.py

```diff
@@ -189,6 +189,31 @@ def __init__(self, label: Optional[str] = None, ctrl_state: Optional[Union[str,
             "cx", 2, [], num_ctrl_qubits=1, label=label, ctrl_state=ctrl_state, base_gate=XGate()
         )
 
+    def _define_qasm3(self):
+        from qiskit.qasm3.ast import (
+            Constant,
+            Identifier,
+            Integer,
+            QuantumBlock,
+            QuantumGateModifier,
+            QuantumGateModifierName,
+            QuantumGateSignature,
+            QuantumGateDefinition,
+            QuantumGateCall,
+        )
+
+        control, target = Identifier("c"), Identifier("t")
+        call = QuantumGateCall(
+            Identifier("U"),
+            [control, target],
+            parameters=[Constant.pi, Integer(0), Constant.pi],
+            modifiers=[QuantumGateModifier(QuantumGateModifierName.ctrl)],
+        )
+        return QuantumGateDefinition(
+            QuantumGateSignature(Identifier("cx"), [control, target]),
+            QuantumBlock([call]),
+        )
```

Comment:
This is pretty temporary: just like the QASM 2 exporter, we probably will want a way for gates to define their QASM 3 representation, but with QASM 3 being far more complex, the structure of exporting now goes "Terra -> AST -> string".  So gates have to return something that can be made into an AST object.  For the time being, while we're vendoring our own AST, we return that.  In the future, once we've shifted to using the reference AST, which crucially will include a "string -> AST" parser, we'll be able to allow gates to return their QASM 3 definition as a string, and in those cases we'll be able to use the parser to raise it to a full AST node.

Having the definitions in proper AST representation is nice because we get a separation of concerns between AST generation and text output, and we can optionally do much more compiler-like things, like verifying that the instruction is valid, and that the signature is consistent with (say) another definition given in an include file.

The CX gate here is mostly just an example - we may want to move this to a different gate, or just define a new test gate in a unit test, since in reality `cx` should be mapped to builtin names most of the time.

Reply:
> The CX gate here is mostly just an example

Not sure. The problem with `cx` is that has no definition in Qiskit. Not having a `_define_qasm3` means that `cx` is opaque when the standard include is missing.

Reply:
That said, I'd be happy if the way we do it is to define `_define_qasm3()` on `ControlledGate`, and have the inherited to all members, but that starts to get slightly more complicated, because then `ControlledGate._define_qasm3` will need to have a way to query the exporter for how the inner gate is defined.